daemon/policy.go

```diff
@@ -220,17 +221,37 @@ func (d *Daemon) PolicyAdd(rules policyAPI.Rules, opts *AddOptions) (uint64, err
 	}
 
 	d.policy.Mutex.Lock()
+
 	// removedPrefixes tracks prefixes that we replace in the rules. It is used
 	// after we release the policy repository lock.
 	var removedPrefixes []*net.IPNet
+
+	// policySelectionWG is used to signal when the updating of all of the
+	// caches of endpoints in the rules which were added / updated have been
+	// updated.
+	var policySelectionWG sync.WaitGroup
+
+	// Get all endpoints at the time rules were added / updated so we can figure
+	// out which endpoints to regenerate / bump policy revision.
+	allEndpoints := endpointmanager.GetEndpoints()
+	epsIdentityConsumers := make([]policy.IdentityConsumer, len(allEndpoints))
+
+	// Need to explicitly convert endpoints to IdentityConsumers.
+	// See: https://github.com/golang/go/wiki/InterfaceSlice
+	for _, ep := range allEndpoints {
+		epsIdentityConsumers = append(epsIdentityConsumers, ep)
+	}
+
+	endpointsToRegen := policy.NewIDSet()
+
 	if opts != nil {
 		if opts.Replace {
 			for _, r := range rules {
 				oldRules := d.policy.SearchRLocked(r.Labels)
 				removedPrefixes = append(removedPrefixes, policy.GetCIDRPrefixes(oldRules)...)
 				if len(oldRules) > 0 {
 					d.dnsRuleGen.StopManageDNSName(oldRules)
-					d.policy.DeleteByLabelsLocked(r.Labels)
+					d.policy.DeleteByLabelsLocked(r.Labels, epsIdentityConsumers, endpointsToRegen, &policySelectionWG)
```

Comment:
Would love to see a `policyMutationContext` be passed in instead of individual arguments. It would clean up the new code a lot.